pkg/vsphere/vsan/synced_cache.go

```diff
@@ -0,0 +1,439 @@
+		return nil
+	}
+
+	if dsType, _ := ds.Type(ctx); dsType != types.HostFileSystemVolumeFileSystemTypeVsan {
+		log.Debugf("datastore %s is not vsan, no need to build vsan dom cache", ds.InventoryPath)
+		return nil
+	}
+
+	hosts, err := ds.AttachedHosts(ctx)
+	if err != nil {
+		err = errors.Errorf("failed to get attached hosts for datastore %s: %s", ds.InventoryPath, err)
+		log.Error(err)
+		return err
+	}
+	if len(hosts) == 0 {
+		err = errors.Errorf("no host attached to datastore %s", ds.InventoryPath)
+		log.Error(err)
+		return err
+	}
+
+	// vsan object can be query and removed from any host whatever the owner is, so here just get the first host
+	hc := hosts[0].ConfigManager()
+	var h mo.HostSystem
+
+	err = hc.Properties(ctx, hc.Reference(), []string{"configManager.vsanInternalSystem"}, &h)
+	if err != nil {
+		err = errors.Errorf("failed to get vsanInternalSystem for host %s: %s", hosts[0].Reference(), err)
+		log.Error(err)
+		return err
+	}
+
+	if h.ConfigManager.VsanInternalSystem == nil {
+		err = errors.Errorf("vsanInternalSystem of host %s is empty", hosts[0].Reference())
+		log.Error(err)
+		return err
+	}
+
+	hvis := NewHostVsanInternalSystem(ds.Client(), *h.ConfigManager.VsanInternalSystem)
+	dsc := &vsanDSDomCache{
+		ds:    ds,
+		hvis:  hvis,
+		uuids: make(map[string]string),
+		paths: make(map[string]string),
+		m:     &sync.Mutex{},
+	}
+	c.m.Lock()
+	defer c.m.Unlock()
+	c.dsMap[ds.Reference().String()] = dsc
+	log.Debugf("Dom cache for datastore %s is created, start to refresh", ds.InventoryPath)
+	c.Refresh(ds.Reference().String(), nil)
+	return nil
+}
+
+func (c *syncedDomCache) getNonRefreshingDS(requests []*dsWaitGroup) []*dsWaitGroup {
+	defer trace.End(trace.Begin(fmt.Sprintf("%#v", requests)))
+
+	c.m.Lock()
+	defer c.m.Unlock()
+	var ret []*dsWaitGroup
+	for i := range requests {
+		if _, ok := c.refreshings[requests[i].dsRef]; ok {
+			// this datastore is refreshing, add wait group into the list
+			if requests[i].wg != nil {
+				log.Debugf("datastore %s is refreshing, append waiting group only", requests[i].dsRef)
+				c.refreshings[requests[i].dsRef] = append(c.refreshings[requests[i].dsRef], requests[i].wg)
+			}
+		} else {
+			ret = append(ret, requests[i])
+		}
+	}
+	return ret
+}
+
+// doRefresh should be called by run only to make sure concurrency is handled properly
+func (c *syncedDomCache) doRefresh(ctx context.Context, requests []*dsWaitGroup) error {
+	defer trace.End(trace.Begin(fmt.Sprintf("%#v", requests)))
+	dsws := c.getNonRefreshingDS(requests)
+	if len(dsws) == 0 {
+		log.Debugf("Requested datastores %#v is refreshing", requests)
+		return nil
+	}
+
+	c.m.Lock()
+	defer c.m.Unlock()
+	var errs []string
+	for i, dsw := range dsws {
+		dsc, ok := c.dsMap[dsw.dsRef]
+		if !ok {
+			err := errors.Errorf("requested datastore %s does not exist", c.dsMap[dsw.dsRef])
+			log.Error(err)
+			errs = append(errs, err.Error())
+			continue
+		}
+		if dsws[i].wg != nil {
```

Comment:
why do you access element by index here, meanwhile you access element as a range provides?